TypeScript 4.8: Error when no type argument is provided for `render`

When invoking `render` with no type argument for rendering context...

```ts
// no-context-test.ts
import { render } from '@ember/test-helpers';
import { hbs } from 'ember-cli-htmlbars';
import { setupRenderingTest } from 'ember-qunit';
import { module, test } from 'qunit';

module('No Context test', function (hooks) {
  setupRenderingTest(hooks);

  test('it renders', async function () {
    await render(hbs`hello`);
  });
});
```

...I'm getting the following error when running `glint` after upgrading to TS 4.8:

```
no-context-test.ts:10:18 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(template: TestTemplate<void>): Promise<void>', gave the following error.
    Argument of type 'new () => Invokable<(args: EmptyObject) => AcceptsBlocks<EmptyObject, null>> & HasContext<TemplateContext<void, EmptyObject, EmptyObject, void>>' is not assignable to parameter of type 'TestTemplate<void>'.
      Construct signature return types 'Invokable<(args: EmptyObject) => AcceptsBlocks<EmptyObject, null>> & HasContext<TemplateContext<void, EmptyObject, EmptyObject, void>>' and 'HasContext<TemplateContext<void, EmptyObject, EmptyObject, null>>' are incompatible.
        The types of '[Context]' are incompatible between these types.
          Type 'TemplateContext<void, EmptyObject, EmptyObject, void>' is not assignable to type 'TemplateContext<void, EmptyObject, EmptyObject, null>'.
            Type 'void' is not assignable to type 'null'.
  Overload 2 of 2, '(template: TemplateFactory): Promise<void>', gave the following error.
    Argument of type 'new () => Invokable<(args: EmptyObject) => AcceptsBlocks<EmptyObject, null>> & HasContext<TemplateContext<void, EmptyObject, EmptyObject, void>>' is not assignable to parameter of type 'TemplateFactory'.

10     await render(hbs`hello`);
                    ~~~~~~~~~~
```

If I add `{}` or `Record<string, never>` as a type argument to `render`, the error goes away.

```ts
await render<{}>(hbs`hello`);
```

Notably, TS 4.8 makes a [breaking change re: unconstrained generics](https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#unconstrained-generics-no-longer-assignable-to).

--- a/packages/environment-ember-loose/-private/dsl/integration-declarations.d.ts
+++ b/packages/environment-ember-loose/-private/dsl/integration-declarations.d.ts
@@ -101,7 +101,7 @@ import '@ember/test-helpers';
 import 'ember-cli-htmlbars';
 
 type TestTemplate<T> = abstract new () => HasContext<
-  TemplateContext<T, EmptyObject, EmptyObject, null>
+  TemplateContext<T, EmptyObject, EmptyObject, void>
 >;
 
 declare module '@ember/test-helpers' {

--- a/test-packages/ts-ember-app/package.json
+++ b/test-packages/ts-ember-app/package.json
@@ -19,8 +19,8 @@
     "lint:js": "eslint . --cache",
     "lint:js:fix": "eslint . --fix",
     "start": "ember serve",
-    "test": "npm-run-all lint test:*",
-    "typecheck": "glint",
+    "test": "npm-run-all test:*",
+    "test:typecheck": "glint",
     "test:tsc": "tsc --noEmit",
     "test:browsers": "ember test"
   },

--- a/test-packages/ts-ember-app/tests/integration/components/render-test.ts
+++ b/test-packages/ts-ember-app/tests/integration/components/render-test.ts
@@ -1,0 +1,82 @@
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import { render } from '@ember/test-helpers';
+import { hbs } from 'ember-cli-htmlbars';
+import type { TestContext } from '@ember/test-helpers';
+
+module('Integration | Component | bar', function (hooks) {
+  setupRenderingTest(hooks);
+
+  module('with no type arg', function () {
+    test('is fine with static content', async function (assert) {
+      await render(hbs`Hello!`);
+
+      assert.dom().hasText('Hello!');
+    });
+
+    test('causes a type error when accessing properties on `this`', async function (assert) {
+      await render(hbs`
+        {{! 
+          @glint-ignore
+          In TS <= 4.7, this is incorrectly allowed because TS infers 'never' as the type of 'this'.
+          In TS 4.8+, this correctly produces an error.
+          TODO: Once we only support 4.8+, this should become a @glint-expect-error.
+        }}
+        Hello, {{this.target}}!
+      `);
+
+      assert.dom().hasText('Hello, !');
+    });
+
+    test('causes a type error when accessing @args', async function (assert) {
+      await render(hbs`
+        {{! @glint-expect-error: no args in test templates }}
+        Hello, {{@target}}!
+      `);
+
+      assert.dom().hasText('Hello, !');
+    });
+  });
+
+  module('with a type arg', function (hooks) {
+    interface MyContext extends TestContext {
+      target: string;
+    }
+
+    hooks.beforeEach(function (this: MyContext) {
+      this.target = 'World';
+    });
+
+    test('is fine with static content', async function (assert) {
+      await render<MyContext>(hbs`Hello!`);
+
+      assert.dom().hasText('Hello!');
+    });
+
+    test('is fine accessing existing properties on `this`', async function (assert) {
+      await render<MyContext>(hbs`
+        Hello, {{this.target}}!
+      `);
+
+      assert.dom().hasText('Hello, World!');
+    });
+
+    test('causes a type error when accessing nonexistent properties on `this`', async function (assert) {
+      await render<MyContext>(hbs`
+        {{! @glint-expect-error: foo isn't a defined property }}
+        Hello, {{this.foo}}!
+      `);
+
+      assert.dom().hasText('Hello, !');
+    });
+
+    test('causes a type error when accessing @args', async function (assert) {
+      await render(hbs`
+        {{! @glint-expect-error: no args in test templates }}
+        Hello, {{@target}}!
+      `);
+
+      assert.dom().hasText('Hello, !');
+    });
+  });
+});

